Figuring out master node by LabelNodeRoleMaster

FengyunPan opened the pull request:
As of 1.6, we will taint the master and recognize nodes labeled
as master, so check the labels of node firstly in IsMasterNode().

**Special notes for your reviewer**:
/assign justinsb
related to #44745

**Release note**:
```release-note
NONE
```

FengyunPan commented:
/assign @smarterclayton

zhouhaibing089 commented on pkg/util/system/system_utils.go:
```diff
@@ -18,14 +18,25 @@ package system
 	"strings"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
```
I do not think it is right to import an application(`kubeadm`) code in a `util` package.

FengyunPan replied:
Yeah, that make sense. But in my PR, I just want to import LabelNodeRoleMaster.
I think your issue is that why LabelNodeRoleMaster comes from an application(kubeadm), why not types.
So if the file of LabelNodeRoleMaster is changed, we will change this import, Do you agree with it?

zhouhaibing089 replied:
ref: https://github.com/kubernetes/kubernetes/pull/39112.

Looking at the comments, it is not going to stay in core soon, 😞

FengyunPan replied:
All right, thank you.
The "moving the new labels and the old ones below to pkg/api/types.go" from that PR, I will update my PR and serviceController.go after merging that PR.

xiangpengzhao commented on pkg/util/system/system_utils_test.go:
```diff
@@ -21,27 +21,89 @@ import (
-		{"foo-master-", false},
-		{"foo-master-a", false},
-		{"foo-master-ab", false},
-		{"foo-master-abc", true},
```
Just curious, why `foo-master-abc` is a true case?
@jszczepkowski as you wrote the original code in #36219, could you please explain this? Thanks!

johnSchnake replied:
Did anyone ever find out why you need `3` here? I have a case where my node is named `foo-master-1` (which seems reasonable) but I see in my logs that this has implications for HPA/metrics gathering (maybe among other things).  Trying to figure out the origins of the `3` @FengyunPan @jszczepkowski @xiangpengzhao

FengyunPan commented on pkg/util/system/system_utils_test.go:
```diff
@@ -21,27 +21,89 @@ import (
-		{"foo-master-", false},
-		{"foo-master-a", false},
-		{"foo-master-ab", false},
-		{"foo-master-abc", true},
```
@xiangpengzhao The nodeName[:len(nodeName)-3] will be 'foo-master-', strings.HasSuffix() just check the string which end with 'master-', so its result is true.

xiangpengzhao replied:
That's exactly what I'm curious with. Why is `3` here? 

```go
 if len(nodeName) >= 10 {
 	return strings.HasSuffix(nodeName[:len(nodeName)-3], "master-")
 }
```

FengyunPan replied:
Yeah, I see. I am also confused about these code. My PR just add a a better way of figuring out if given node is a registered master and save the original code.

justinsb commented:
Ouch - this is plainly a much better heuristic than the heuristic we're using currently!

/lgtm

FengyunPan commented:
/ok-to-test

FengyunPan commented:
/test pull-kubernetes-e2e-kops-aws

smarterclayton commented:
@liggitt re: use of master node labels

smarterclayton commented on pkg/util/system/BUILD:
```diff
@@ -9,13 +9,18 @@ load(
     srcs = ["system_utils_test.go"],
     library = ":go_default_library",
     deps = [
+        "//cmd/kubeadm/app/constants:go_default_library",
```
Yeah, won't be allowed.  You'll have to duplicate the constant.

FengyunPan replied:
Done, thank you.

smarterclayton commented on pkg/util/system/system_utils.go:
```diff
@@ -18,14 +18,24 @@ package system
 // TODO: find a better way of figuring out if given node is a registered master.
-func IsMasterNode(nodeName string) bool {
+// IsMasterNode returns true if given node is a registered master.
+func IsMasterNode(node v1.Node) bool {
```
If this is only used in tests it doesn't belong in pkg/util/system.  Move it to test/e2e/...

smarterclayton replied:
As a general rule, a heuristic like this may not be in the core code base.  It may only be layered on top.

FengyunPan replied:
Currently, caller just pass a node name and IsMasterNode() check it by string.
IsMasterNode() of my patch need the labels of nodes, so I use v1.node instead of string.

timothysc replied:
@FengyunPan could you move this to test/utils and we can proceed forwards here.

justinsb replied:
It's used in `pkg/controller/node/node_controller.go`

luxas commented:
cc @kubernetes/sig-cluster-lifecycle-pr-reviews 
/assign @timothysc 

This is pretty much related to conformance test running -- I've seen cases where features have been skipped because I don't have a "master" node in the cluster...

liggitt commented:
echoing what @smarterclayton said:

> As a general rule, a heuristic like this may not be in the core code base. It may only be layered on top. Tests are considered layered on top. If you use this in a conformance test, then the heuristic must work even if the label is not present (conformance tests may not dictate use of these labels).

FengyunPan2 committed:
Figuring out master node by LabelNodeRoleMaster

As of 1.6, we will taint the master and recognize nodes labeled
as master, so check the labels of node firstly in IsMasterNode().

FengyunPan commented:
@liggitt @smarterclayton
The 'IsMasterNode(node v1.Node)' is not only used in tests, this PR check node.Label by v1.node.
And I move "LabelNodeRoleMaster = "node-role.kubernetes.io/master"" within 'pkg/util/system/system_utils.go'.

smarterclayton commented:
> The 'IsMasterNode(node v1.Node)' is not only used in tests, this PR check node.Label by v1.node.

As I mentioned before, the core code base cannot rely on these labels - they are convention only.  Putting them in pkg/util/system violates that guideline.

These labels are allowed to be used by higher level systems, but we do not require or dictate their use within Kubernetes/Kubernetes

FengyunPan commented:
@smarterclayton I see, thank for your comment.
If I can't use this label 'LabelNodeRoleMaster', my PR is broken.
So do you have better idea to help this PR?

justinsb commented:
The current heuristic is that node names must end with "master" or must end in `master-XXXX` (if I read it correctly).

I think this PR is a great improvement to that heuristic - I suspect that only works on GCE.
